rubble/src/att/mod.rs

```diff
@@ -35,47 +35,69 @@ mod server;
 mod uuid;
 
 use self::{handle::*, pdus::*};
-use crate::{utils::HexSlice, Error};
+use crate::Error;
 
 pub use self::handle::{Handle, HandleRange};
 pub use self::server::{AttributeServer, AttributeServerTx};
 pub use self::uuid::AttUuid;
 
+/// An attribute value that can be represented as a byte slice.
+pub trait AttrValue {
+    fn as_slice(&self) -> &[u8];
+}
+
+impl AttrValue for &[u8] {
+    fn as_slice(&self) -> &[u8] {
+        self
+    }
+}
+
+impl AttrValue for () {
+    fn as_slice(&self) -> &[u8] {
+        &[]
+    }
+}
+
 /// An ATT server attribute
-pub struct Attribute<'a> {
+pub struct Attribute<T>
+where
+    T: AttrValue,
+{
     /// The type of the attribute as a UUID16, EG "Primary Service" or "Anaerobic Heart Rate Lower Limit"
     pub att_type: AttUuid,
     /// Unique server-side identifer for attribute
     pub handle: Handle,
     /// Attribute values can be any fixed length or variable length octet array, which if too large
     /// can be sent across multiple PDUs
-    pub value: HexSlice<&'a [u8]>,
+    pub value: T,
 }
 
-impl<'a> Attribute<'a> {
+impl<T: AttrValue> Attribute<T> {
     /// Creates a new attribute.
-    pub fn new(att_type: AttUuid, handle: Handle, value: &'a [u8]) -> Self {
+    pub fn new(att_type: AttUuid, handle: Handle, value: T) -> Self {
         assert_ne!(handle, Handle::NULL);
         Attribute {
             att_type,
             handle,
-            value: HexSlice(value),
+            value: value,
         }
     }
 
     /// Retrieves the attribute's value as a slice.
-    pub fn value(&self) -> &'a [u8] {
-        self.value.as_ref()
+    pub fn value(&self) -> &[u8] {
+        self.value.as_slice()
     }
 
     /// Overrides the previously set attribute's value.
-    pub fn set_value(&mut self, value: &'a [u8]) {
-        self.value = HexSlice(value)
+    pub fn set_value(&mut self, value: T) {
+        self.value = value;
     }
 }
 
 /// Trait for attribute sets that can be hosted by an `AttributeServer`.
 pub trait AttributeProvider {
+    type ValueType: AttrValue;
```

Comment:
Ah, I see the problem here. You should be able to downcast `&Attribute<T>` to `&Attribute<dyn AttrValue>` by adding a `?Sized` bound. If you then make all `AttributeProvider` method accept *that* you can arbitrarily mix `AttrValue` types.

Reply:
Right, that is certainly doable. I was thinking that using dyn does have an impact of indirection vs. using a fixed type, and I wasn't sure which approach you preferred. If you're ok with that, then I will change it to use "trait objects" as I think its called. I think having the flexibility is beneficial here, and the cost should be minimal I guess.

Reply:
Yes, I don't expect this to have a tangible impact, except perhaps on code size when using a lot of different value types. But that can always be fixed later.

Reply:
Updated the PR with the change now.